cpp/test/sparse/preprocess_coo.cu

```diff
@@ -0,0 +1,214 @@
+      raft::make_device_vector<Type_f, int64_t>(handle, c_matrix.structure_view().get_n_rows() * k);
+
+    raft::sparse::neighbors::brute_force_knn<Index_, Type_f>(c_matrix,
+                                                             c_matrix,
+                                                             out_indices.data_handle(),
+                                                             out_dists.data_handle(),
+                                                             k,
+                                                             handle,
+                                                             c_matrix.structure_view().get_n_rows(),
+                                                             c_matrix.structure_view().get_n_rows(),
+                                                             raft::distance::DistanceType::L1);
+
+    if (bm25_on) {
+      ASSERT_TRUE(raft::devArrMatch<Index_>(out_idxs_bm25.data_handle(),
+                                            out_indices.data_handle(),
+                                            out_indices.size(),
+                                            raft::Compare<Index_>(),
+                                            stream));
+      ASSERT_TRUE(raft::devArrMatch<Type_f>(out_dists_bm25.data_handle(),
+                                            out_dists.data_handle(),
+                                            out_dists.size(),
+                                            raft::CompareApprox<Type_f>(2e-5),
+                                            stream));
+    } else {
+      ASSERT_TRUE(raft::devArrMatch<Index_>(out_idxs_tfidf.data_handle(),
+                                            out_indices.data_handle(),
+                                            out_indices.size(),
+                                            raft::Compare<Index_>(),
+                                            stream));
+      ASSERT_TRUE(raft::devArrMatch<Type_f>(out_dists_tfidf.data_handle(),
+                                            out_dists.data_handle(),
+                                            out_dists.size(),
+                                            raft::CompareApprox<Type_f>(2e-5),
+                                            stream));
+    }
+    RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
+  }
+
+ protected:
+  raft::resources handle;
+  cudaStream_t stream;
+
+  SparsePreprocessInputs<Type_f, Index_> params;
+};
+
+using SparsePreprocessTfidfCoo = SparsePreprocessCoo<float, int>;
+TEST_P(SparsePreprocessTfidfCoo, Result) { Run(false); }
+
+using SparsePreprocessBm25Coo = SparsePreprocessCoo<float, int>;
+TEST_P(SparsePreprocessBm25Coo, Result) { Run(true); }
+
+const std::vector<SparsePreprocessInputs<float, int>> sparse_preprocess_inputs = {
+  {12,                                                   // n_rows
+   5,                                                    // n_cols
+   {0, 3, 4, 5, 6, 7, 8, 9, 10, 11},                     // rows
```

Comment:
We probably want to test with larger data. One thing we often do in tests is use a data generator (I would suggest RMAT here) and we'll write kernels that evaluate the naive solution right in the gtest for us to compare against. tfidf and bm25 are straightforward enought hat I think writing naive kernels would be the way to go.

Reply:
Just to make sure everyone is on the same page, shifting over to rmat generation will mean we do not really have control over size of the nnz edges, because there are possible duplicate edges.